src/librustc_builtin_macros/asm.rs

```diff
@@ -264,27 +270,72 @@ fn parse_args<'a>(
     Ok(args)
 }
 
+fn warn_duplicate_option<'a>(p: &mut Parser<'a>, span: Span) {
+    let mut warn = if let Ok(snippet) = p.sess.source_map().span_to_snippet(span) {
+        p.sess
+            .span_diagnostic
+            .struct_span_warn(span, &format!("the `{}` option was already provided", snippet))
+    } else {
+        p.sess.span_diagnostic.struct_span_warn(span, "this option was already provided")
+    };
+    warn.span_suggestion(
+        span,
+        "remove this option",
+        String::new(),
+        Applicability::MachineApplicable,
+    );
+    warn.emit();
+}
+
 fn parse_options<'a>(p: &mut Parser<'a>, args: &mut AsmArgs) -> Result<(), DiagnosticBuilder<'a>> {
     let span_start = p.prev_token.span;
 
     p.expect(&token::OpenDelim(token::DelimToken::Paren))?;
 
     while !p.eat(&token::CloseDelim(token::DelimToken::Paren)) {
         if p.eat(&token::Ident(sym::pure, false)) {
-            args.options |= ast::InlineAsmOptions::PURE;
+            if !args.option_is_set(ast::InlineAsmOptions::PURE) {
+                args.options |= ast::InlineAsmOptions::PURE;
+            } else {
+                warn_duplicate_option(p, p.prev_token.span);
+            }
```

Comment:
There's a lot of duplicated code here, maybe you should push more logic into a help function:
```rust
fn try_parse_option(&mut self, span: Span, sym: Symbol, opt: InlineAsmOptions) -> bool
```

Reply:
Do you mean something like:

```rust
fn try_parse_option<'a>(
    p: &mut Parser<'a>,
    args: &mut AsmArgs,
    symbol: Symbol,
    option: ast::InlineAsmOptions,
) -> bool {
    if p.eat(&token::Ident(symbol, false)) {
        if !args.option_is_set(option) {
            args.options |= option;
        } else {
            err_duplicate_option(p, p.prev_token.span);
        }
        true
    } else {
        false
    }
}
```

that's used like:

```rust
if try_parse_option(p, args, sym::pure, ast::InlineAsmOptions::PURE) {
} else if try_parse_option(p, args, sym::nomem, ast::InlineAsmOptions::NOMEM) {
} /* ... */
```

Reply:
If that's what you mean, how should I handle the last case? Currently, it's an `else` with `p.expect(&token::Ident(sym::att_syntax, false))?;` inside of it

Reply:
Ah good point, maybe keep that part out of the helper function. The function can just set the option and emit an error if it was already set. I don't think it is necessary to have separate functions for `option_is_set` and `err_duplicate_opiton`.

Reply:
Although I think those separate functions make the source code more readable